### ISSUE TITLE
[Bug]: Vehicle Preview glitches out and actually spawns the vehicles

### ISSUE THREAD

**Issue by GreySideMM** [Yimura/YimMenu#2675]
### Describe the bug

Only / at least at one certain spot on the map the preview glitches out. Instead of seeing the cars nothing shows up. The cars are spawned apparently underneath the map. Only after like 10-20sec suddenly usually dozens of vehicles start showing up. Apparently being teleported to the surface again after they fell into the void beneath the map. The vehicles are spawned as persistent, they will remain there forever. They also are networked.
Closely related to #2077.

### Steps To Reproduce

1. Create a new or join a existing session. (New one seems to be easier to reproduce.)
2. Teleport to the LS Customs south of the airfield at Sandy Shores.
3. Be on the street, face towards LSC. (The rough area should be good enough.)
4. You have to spawn at least a car once. Maybe do this several times.
5. Keep moving the mouse above multiple entries in the vehicle list to trigger a new/different preview.
6. You should now notice that no longer previews appear. Wait up to half a minute or try again after walking some meters and restart at 4.
7. Enjoy the piling up of docens of cars.

### Expected Behavior

No spawned cars. No networking of these cars.

### Logs and/or screenshots

```
<paste your logs here>
```
![20240105092332_1](https://github.com/YimMenu/YimMenu/assets/155721485/0ae6b499-cc17-4b48-bf83-c59bd3133f1b)
![20240105092248_1](https://github.com/YimMenu/YimMenu/assets/155721485/7a0fe8c8-ae20-4967-a265-15d980c82818)

### Store

Steam

### Language

English

### YimMenu Language

English (US)

### Build

- [X] I've confirmed the issue exists on the latest version of YimMenu

### Coload

- [X] I've tested this without co-loading anything.

**Comment by maybegreat48** [comment]
Does the game do weird stuff after this happens? Like being unable to play missions or the game eventually crashing?

**Comment by gir489returns** [comment]
@Yimura Should we fix this? Seems like a super non-issue.

**Comment by GreySideMM** [comment]
> Does the game do weird stuff after this happens? Like being unable to play missions or the game eventually crashing?
The game does not crash from this. I can't say if or if not missions are playable after this, would need to test this. Freemode Missions might be effected, as at least I get an error from the menu that it can't spawn no more vehicles. So area near by might be effected? (Idk how this works. Spawning worked later sometimes again, sometimes not, iirc.)
Any specific mission in mind?

**Comment by Yimura** [comment]
> @Yimura Should we fix this? Seems like a super non-issue.

The issue is definitely there but it seems to be a bug in the game, there's one thing we can try to fix this however if that doesn't work it's a lost cause.

**Comment by roadsport** [comment]
This seems to be way more common since the last Yim update. If it happens once then it no longer wants to spawn vehicles at all unless you re-inject the menu

**Comment by gir489returns** [comment]
> This seems to be way more common since the last Yim update. If it happens once then it no longer wants to spawn vehicles at all unless you re-inject the menu

We know about it, and can replicate it, it's just we don't really care to fix it. We suspect it has to do with the fact it's jamming the Vehicle pool with crap until it overflows and we can't spawn anymore cars. We use a singleton instance of the vehicle, and make sure to clean up the vehicle when the user's cursor falls out of scope of the highlighted vehicle, but that sems to be not enough.

**Comment by gir489returns** [comment]
@GreySideMM Can you please test https://github.com/YimMenu/YimMenu/pull/2879 to make sure I patched it properly?

**Comment by GreySideMM** [comment]
@gir489returns I tested it now and I'm sorry to say but it's now kind of even worse than before: The vehicles glitch spawn now way faster than before (and also more).

**Comment by gir489returns** [comment]
> @gir489returns I tested it now and I'm sorry to say but it's now kind of even worse than before: The vehicles glitch spawn now way faster than before (and also more).

Did you test the build from release? I just did a rapid scroll of the list, and no extraneous cars were spawned.

**Comment by GreySideMM** [comment]
I did test my own version (up to date, with RelWithDebInfo) as well as the binary from #2879 (which I now figured was not up to date). You have been on the right spot?
I think this might be important here: I don't have such a high framerate, usually ~65 or ~35 fps. I noticed the yield()s and some time based changes.

**Comment by gir489returns** [comment]
> I did test my own version (up to date, with RelWithDebInfo) as well as the binary from #2879 (which I now figured was not up to date). You have been on the right spot? I think this might be important here: I don't have such a high framerate, usually ~65 or ~35 fps. I noticed the yield()s and some time based changes.

Yes I noticed that too. It also seems to be exacerbated online. I only tested it in SP. I’m going to rewrite it from scratch and use mutexes and a cooldown. It won’t dispatch a new preview request until the last one was > 200MS. The idea being if you’re going up and down rapidly on the list, it’ll wait until you have stopped and have your cursor hovering over something before kicking off a new preview request instead of trying to do 500 of them.

**Comment by GreySideMM** [comment]
> Yes I noticed that too. It also seems to be exacerbated online. I only tested it in SP. I’m going to rewrite it from scratch and use mutexes and a cooldown. It won’t dispatch a new preview request until the last one was > 200MS. The idea being if you’re going up and down rapidly on the list, it’ll wait until you have stopped and have your cursor hovering over something before kicking off a new preview request instead of trying to do 500 of them.

Ah ok. You probably should implement the cooldown first, could be enough to fix this.
With the newer version I also noticed that now you can already hear the vehicles explode (and collide, I think) even several seconds before the first vehicles get visible at all. So they might spawn within each other (somewhere below the map). (I assumed prior this might just be a position problem, as the preview is not even visible for a single frame.)
I only tested online, but usually alone. Just fyi.

**Comment by Yimura** [comment]
The vehicles are exploding under the map I believe (that's why hear them but don't see them).

**Comment by gir489returns** [comment]
@GreySideMM I have redesigned the preview system entirely. Please try the build from PR #2888. Thanks in advanced.

**Comment by gir489returns** [comment]
This seems to be a larger issue of us exhausting the vehicle_pool, then the preview system getting stuck in an infinite loop trying to request vehicles and then getting stuck, only for those vehicles to later pop into existence.
